TELCODOCS-1914 Add troubleshooting information to MetalLB 2 (#107634)
openshift/openshift-docs: modules/nw-metallb-troubleshoot-bgp.adoc

.Procedure

> . In {product-title} 4.17 and later, MetalLB delegates BGP to a separate FRR-K8s daemon set running in the `openshift-frr-k8s` namespace. The speaker pod itself no longer contains an FRR container. Display the names of the `frr-k8s` pods by running the following command:

Reviewer: delegates == uses the frr-k8s daemon which is deployed by CNO .
That means if there is an issue e.g. daemon set of frr-k8s this is a fix on CNO

service integrated-vtysh-config
!
router bgp 64501
 no bgp ebgp-requires-policy
 no bgp default ipv4-unicast
 bgp graceful-restart preserve-fw-state
 no bgp network import-check
 neighbor 192.168.122.12 remote-as 64500
 !
 address-family ipv4 unicast
  network 192.168.122.210/32
  neighbor 192.168.122.12 activate
  neighbor 192.168.122.12 route-map 192.168.122.12-in in
  neighbor 192.168.122.12 route-map 192.168.122.12-out out
 exit-address-family
exit
!
ip prefix-list 192.168.122.12-inpl-ipv4 seq 1 deny any
ip prefix-list 192.168.122.12-allowed-ipv4 seq 1 permit 192.168.122.210/32
!
ipv6 prefix-list 192.168.122.12-allowed-ipv6 seq 1 deny any
ipv6 prefix-list 192.168.122.12-inpl-ipv4 seq 2 deny any
!
route-map 192.168.122.12-out permit 1
 match ip address prefix-list 192.168.122.12-allowed-ipv4
exit
!
route-map 192.168.122.12-out permit 2
 match ipv6 address prefix-list 192.168.122.12-allowed-ipv6
exit
!
route-map 192.168.122.12-in permit 3
 match ip address prefix-list 192.168.122.12-inpl-ipv4
exit
!
route-map 192.168.122.12-in permit 4
 match ipv6 address prefix-list 192.168.122.12-inpl-ipv4
exit
!
ip nht resolve-via-default
!
ipv6 nht resolve-via-default
!
end
----
+
where:
+

> * `router bgp 64501`:: This is the local Autonomous System Number (ASN) for your MetalLB speakers.

Reviewer 2: There's something going wrong here in the preview
I think it's to do with mixing `*` and `::` in one “where” block 
If you check the [preview](https://107634--ocpdocs-pr.netlify.app/openshift-enterprise/latest/networking/ingress_load_balancing/metallb/metallb-troubleshoot-support.html#nw-metallb-troubleshoot-bgp_metallb-troubleshoot-support) you'll see that the "::" is showing 

Seems you need to either pick bullets + `--` or definition list + `::`, not both.

further down on line 161 you've done it differently and that seems to work

* `neighbor 192.168.122.12 remote-as 64500`:: This identifies the external BGP Peer. Specifies that a `neighbor <ip-address> remote-as <peer-ASN>` line exists for each BGP peer custom resource that you added.
** The local ASN is 64501.
** The remote ASN is 64500.

> * `network 192.168.122.210/32`:: This is a specific LoadBalancer IP from your IPAddressPool. It is being advertised as a /32 (a single host route), which is standard for MetalLB.

Reviewer 3: ```suggestion
* `network 192.168.122.210/32`:: This is a specific LoadBalancer IP from your IPAddressPool. It is being advertised as a `/32` (a single host route), which is standard for MetalLB.
```

+
If BFD is enabled for the BGP peer, the output includes additional `bfd` lines under the neighbor configuration and a `bfd` section at the end. The following example shows the relevant differences:
+
.Example output with BFD enabled
----
router bgp 64501
 ...
 neighbor 192.168.122.12 remote-as 64500
 neighbor 192.168.122.12 bfd
 neighbor 192.168.122.12 bfd profile bfd-profile
 ...
exit
!
...
!
bfd
 profile bfd-profile
  minimum-ttl 1
 exit
 !
exit
!
end
----

. Display the BGP summary by running the following command:
+
[source,terminal]
----
$ oc exec -n openshift-frr-k8s <frr-k8s-pod> -c frr -- vtysh -c "show bgp summary"
----
+
.Example output
----

IPv4 Unicast Summary (VRF default):
BGP router identifier 192.168.122.12, local AS number 64501 vrf-id 0
BGP table version 1
RIB entries 1, using 192 bytes of memory
Peers 1, using 725 KiB of memory

Neighbor        V         AS   MsgRcvd   MsgSent   TblVer  InQ OutQ  Up/Down State/PfxRcd   PfxSnt Desc
192.168.122.12  4      64500        37        38        0    0    0 00:32:12            0        1 N/A

Total number of neighbors 1
----
+
where:
+

> `BGP router identifier 192.168.122.12`:: This is the "ID" of your node. Usually, it's the IP of the primary interface.

Reviewer 3: ```suggestion
`BGP router identifier 192.168.122.12`:: This is the "ID" of your node. Usually, it s the IP of the primary interface.
```

Also, why is ID quoted here?

`local AS number 64501`:: This is the ASN you assigned to MetalLB in your `BGPPeer` or `MetalLB` CR.
`Neighbor`:: The IP `192.168.122.12` of your external router (the "Peer").
`AS`:: The ASN of the external router (the "Peer"), which should match the `remote-as` value in your BGP configuration.

> `Up/Down`:: The session has been stable for 32 minutes.

Reviewer 3: ```suggestion
`Up/Down`:: Informs that the session has been stable for 32 minutes.
```
Or `Shows`, `Indicates`, etc.

> `State/PfxRcd`:: This shows the session is established since it's a number, but you have received 0 routes from the peer. Seeing 0 prefixes received is perfectly normal for a standard MetalLB deployment.

Reviewer 3: ```suggestion
`State/PfxRcd`:: This shows the session is established since it is a number, but you have received 0 routes from the peer. Seeing 0 prefixes received is perfectly normal for a standard MetalLB deployment.
```

> `PfxSnt`:: This shows that you have successfully advertised 1 route the LoadBalancer IP to the peer. This confirms MetalLB is doing its job. It has taken one LoadBalancer service IP and successfully told the external router: "If you want to reach this IP, send the traffic to me."

Reviewer 3: ```suggestion
`PfxSnt`:: This shows that you have successfully advertised one route the LoadBalancer IP to the peer. This confirms MetalLB is doing its job. It has taken one LoadBalancer service IP and successfully told the external router: "If you want to reach this IP, send the traffic to me."
```

. Display the BGP peers that received an address pool by running the following command:
+
[source,terminal]
----
$ oc exec -n openshift-frr-k8s <frr-k8s-pod> -c frr -- vtysh -c "show bgp ipv4 unicast"
----
+
Replace `ipv4` with `ipv6` to display the BGP peers that received an IPv6 address pool.
+
.Example output
----
BGP table version is 1, local router ID is 192.168.122.12, vrf id 0
Default local pref 100, local AS 64501
Status codes:  s suppressed, d damped, h history, * valid, > best, = multipath,
               i internal, r RIB-failure, S Stale, R Removed
Nexthop codes: @NNN nexthop's vrf id, < announce-nh-self
Origin codes:  i - IGP, e - EGP, ? - incomplete
RPKI validation codes: V valid, I invalid, N Not found

    Network          Next Hop            Metric LocPrf Weight Path
 *> 192.168.122.210/32
                    0.0.0.0                  0         32768 i

Displayed  1 routes and 1 total paths
----
+
where:
+
`local router ID 192.168.122.12`:: The unique identifier for this specific OpenShift node in the BGP topology.
`local AS 64501`:: The Private Autonomous System Number (ASN) assigned to your MetalLB deployment.
+

> The bottom section is the actual "map" being sent to your upstream router:

Reviewer 2: "The bottom section" is vague here. The way the preview indents this makes it look as if line 202 is a continuation of the preceding bullet point on `local AS 64501` which is not the case.
The way this information is presented half bulleted list, half table isn't very coherent. Id suggest continuing the bulleted list - either with an enry for each element or just one umbrella heading of "BGP route entry fields" or "map data" or something like that. It's awkward information to fit in but this approach isn't very clear.

+
.BGP Route Entry Field Descriptions
[cols="1,2,3", options="header"]
|===
|Field/Symbol |Value in Output |Description and Troubleshooting Impact

|Status Codes
|*>
a|* `*` (Valid): The route is technically correct.
* `>` (Best): The route selected for advertisement. MetalLB only sends "Best" paths to peers. If `>` is missing, the peer will not receive the route.

|Network
|192.168.122.210/32

> |The specific IP assigned to the LoadBalancer service. The /32 mask indicates a host route, ensuring traffic for this specific IP is attracted to this node.

Reviewer 3: ```suggestion
|The specific IP assigned to the LoadBalancer service. The `/32` mask indicates a host route, ensuring traffic for this specific IP is attracted to this node.
```

|Next Hop
|0.0.0.0
|Indicates the route is local to the node. In MetalLB, 0.0.0.0 means the current node is the egress point for this service.

|Metric
|0

> |The Multi-Exit Discriminator (MED). Used to suggest a preferred path to external neighbors. Default is 0.

Reviewer 3: ```suggestion
|The Multi-Exit Discriminator (MED). Used to suggest a preferred path to external neighbors. Default is `0`.
```

|LocPrf
|100

> |Local Preference. Used to prefer an exit point for the entire AS. Default is 100.

Reviewer 3: ```suggestion
|Local Preference. Used to prefer an exit point for the entire AS. Default is `100`.
```

|Weight
|32768

> |An internal FRR priority value. Locally injected routes default to 32768, ensuring the node prefers its own local service path over learned routes.

Reviewer 3: ```suggestion
|An internal FRR priority value. Locally injected routes default to `32768`, ensuring the node prefers its own local service path over learned routes.
```

|Path
|`i`
|The Origin code. `i` (IGP) signifies the route was originated internally through the MetalLB speaker injecting the `IPAddressPool` into the FRR stack.
|===
+
If the table is empty, check if the Service has an IP assigned using the `oc get svc` command. Verify that a `BGPAdvertisement` exists and that its `nodeSelector` or `labelSelector` matches the service and the node you are running the command on.

> If the Next Hop is not `0.0.0.0`, the node might be trying to forward the traffic elsewhere before it even reaches the pods, which can indicate a complex BGP peering issue or an issue with the node's underlying routing table.

Reviewer 3: ```suggestion
If the Next Hop is not `0.0.0.0`, the node might be trying to forward the traffic elsewhere before it even reaches the pods, which can indicate a complex BGP peering issue or an issue with the node underlying routing table.
```

Inanimate object.
